bin/user_config_converter.py

```diff
@@ -0,0 +1,166 @@
+        logger.error(f'in_file "{in_file}" not found.')
+        exit(1)
+
+    if in_file.suffix.lower() == '.json':
+        direction = ConversionDirection.TO_BINARY
+    elif in_file.suffix.lower() == '.p1log':
+        direction = ConversionDirection.TO_JSON
+    else:
+        logger.error(f'in_file has unknown file extension "{in_file.suffix}". Only "*.json" or "*.p1log" are accepted.')
+        exit(1)
+
+    if out_file.is_dir():
+        if direction == ConversionDirection.TO_BINARY:
+            extension = '.p1log'
+        else:
+            extension = '.json'
+
+        out_file = out_file / (in_file.stem + extension)
+
+    if direction == ConversionDirection.TO_BINARY:
+        with open(in_file, 'r') as fd:
+            try:
+                json_data = json.load(fd)
+            except Exception as e:
+                logger.error(f'Could not parse {in_file}: {e}')
+                exit(1)
+
+        if '__version' not in json_data:
+            logger.error(f'{in_file} is missing required field "__version"')
+            exit(1)
+        version_str = json_data['__version']
+        if '__platform_id' not in json_data:
+            logger.error(f'{in_file} is missing required field "__platform_id"')
+            exit(1)
+        platform_id = json_data['__platform_id']
+    else:
+        with open(in_file, 'rb') as fd:
+            binary_data = fd.read()
+        decoder = FusionEngineDecoder()
+        messages = decoder.on_data(binary_data)
+        if len(messages) != 1:
+            logger.error(f'{in_file} did not contain binary FusionEngine data.')
+            exit(1)
+
+        payload = messages[0][1]
+        if not isinstance(payload, PlatformStorageDataMessage):
+            logger.error(f'{in_file} did not contain PlatformStorageDataMessage.')
+            exit(1)
+
+        version_str = str(payload.data_version)
+        platform_id = str(payload.flags)
+        if platform_id == 0:
+            logger.error(f'{in_file} did not specify a platform ID. Device firmware not supported by this tool.')
+            exit(1)
+
+    user_config_dir = Path(repo_root) / 'user_config_loaders' / \
+        f'platform_{platform_id}' / f'version_{version_str.replace(".", "_")}'
+    if not user_config_dir.exists():
+        logger.error(
+            f'The UserConfig for Platform ID {platform_id} and version {version_str} is not known ({user_config_dir} not found). A newer p1-host-tools release may have added support.')
+        exit(1)
+
+    logger.info(f'Loading UserConfig for Platform ID {platform_id} and version {version_str}.')
+
+    # Import the UserConfig class.
+    sys.path.insert(0, str(user_config_dir))
+    module = importlib.import_module(f'user_config_loader.user_config_loader', 'user_config_loader')
+    logger.info(f'Loaded UserConfig version {module.UserConfig.get_version()}.')
+    UserConfig: Type[UserConfigTypingClass] = module.UserConfig
+
+    if direction == ConversionDirection.TO_BINARY:
```

Comment:
Suggested comment for readability: "# Convert to binary or JSON and write the result to disk."